cookbooks/CPO_induced_anisotropic_viscosity/plugin/cpo_induced_anisotropic_viscosity.cc

```diff
@@ -0,0 +1,587 @@
+      return rot_matrix;
+    }
+
+
+
+    template <>
+    void
+    CPO_AV_3D<2>::evaluate (const MaterialModel::MaterialModelInputs<2> &,
+                            MaterialModel::MaterialModelOutputs<2> &) const
+    {
+      Assert (false, ExcNotImplemented());
+    }
+
+
+    template <>
+    void
+    CPO_AV_3D<3>::evaluate (const MaterialModel::MaterialModelInputs<3> &in,
+                            MaterialModel::MaterialModelOutputs<3> &out) const
+    {
+      const int dim=3;
+      const std::shared_ptr<MaterialModel::AnisotropicViscosity<dim>> anisotropic_viscosity =
+        out.template get_additional_output_object<MaterialModel::AnisotropicViscosity<dim>>();
+      EquationOfStateOutputs<dim> eos_outputs (1);
+
+      for (unsigned int q=0; q<in.n_evaluation_points(); ++q)
+        {
+          // Change these according to diffusion dislocation material model I guess
+          equation_of_state.evaluate(in, q, eos_outputs);
+
+          // Get parameters for compute the effective viscosity
+          const std::vector<double> composition = in.composition[q];
+          const std::vector<double> volume_fractions = MaterialUtilities::compute_only_composition_fractions(composition,
+                                                       this->introspection().chemical_composition_field_indices());
+
+          out.densities[q] = eos_outputs.densities[0];
+          out.viscosities[q] = eta;
+          out.thermal_expansion_coefficients[q] = eos_outputs.thermal_expansion_coefficients[0];
+          out.specific_heat[q] = eos_outputs.specific_heat_capacities[0];
+          out.thermal_conductivities[q] = 1;
+          out.compressibilities[q] = eos_outputs.compressibilities[0];
+          out.entropy_derivative_pressure[q] = eos_outputs.entropy_derivative_pressure[0];
+          out.entropy_derivative_temperature[q] = eos_outputs.entropy_derivative_temperature[0];
+
+          // Calculate effective viscosity
+          const SymmetricTensor<2,dim> strain_rate = in.strain_rate[q];
+          const SymmetricTensor<2,dim> deviatoric_strain_rate
+            = (this->get_material_model().is_compressible()
+               ?
+               strain_rate - 1./3. * trace(strain_rate) * unit_symmetric_tensor<dim>()
+               :
+               strain_rate);
+
+          // The computation of the viscosity tensor is only necessary after the simulator has been initialized
+          // and when the condition allows dislocation creep
+          if  ((this->simulator_is_past_initialization()) && (this->get_timestep_number() > 0) && (in.temperature[q]>1000) && (isfinite(determinant(deviatoric_strain_rate))))
+            {
+              const unsigned int ind_vis = this->introspection().compositional_index_for_name("scalar_vis");
+
+              // Create constant value to use for AV
+              const double A_o = 1.1e5*exp(-530000/(8.314*in.temperature[q]));
```

Comment:
please prefix all mathematical functions from the namespace `std` with their namespace, i.e.:

```suggestion
              const double A_o = 1.1e5*std::exp(-530000/(8.314*in.temperature[q]));
```

This is because deal.II or ASPECT could introduce their own `exp` functions in their namespaces, while not expecting anyone to automatically use them, but your code would immediately use them.